HLTrigger/JetMET/plugins/HLTJetHFCleaner.h

```diff
@@ -0,0 +1,221 @@
+      dphiJetMetMin_(iConfig.getParameter<double>("dphiJetMetMin")),
+      jetEtaMin_(iConfig.getParameter<double>("jetEtaMin")),
+      jetEtaMax_(iConfig.getParameter<double>("jetEtaMax")),
+      sigmaEtaPhiDiffMax_(iConfig.getParameter<double>("sigmaEtaPhiDiffMax")),
+      centralEtaStripSizeMax_(iConfig.getParameter<int>("centralEtaStripSizeMax")),
+      cornerCutSigmaEtaEta_(iConfig.getParameter<double>("cornerCutSigmaEtaEta")),
+      cornerCutSigmaPhiPhi_(iConfig.getParameter<double>("cornerCutSigmaPhiPhi")),
+      applySigmaEtaPhiCornerCut_(iConfig.getParameter<bool>("applySigmaEtaPhiCornerCut")),
+      applySigmaEtaPhiCut_(iConfig.getParameter<bool>("applySigmaEtaPhiCut")),
+      applyStripSizeCut_(iConfig.getParameter<bool>("applyStripSizeCut")) {
+  produces<JetCollection>();
+}
+
+template <typename JetType>
+void HLTJetHFCleaner<JetType>::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetup&) const {
+  auto const jets = iEvent.getHandle(jetToken_);
+  auto const& mets = iEvent.get(metToken_);
+
+  // The set of cleaned jets
+  auto cleaned_jets = std::make_unique<JetCollection>();
+  cleaned_jets->reserve(jets->size());
+
+  // Check if length of the MET vector is 1, throw exception otherwise
+  if (mets.size() != 1) {
+    throw cms::Exception("ComparisonFailure") << "Size of reco::MET collection is different from 1";
+  }
+  auto const met_phi = mets.front().phi();
+
+  auto const& sigmaEtaEtas = iEvent.get(sigmaEtaEtaToken_);
+  auto const& sigmaPhiPhis = iEvent.get(sigmaPhiPhiToken_);
+  auto const& centralEtaStripSizes = iEvent.get(centralEtaStripSizeToken_);
+
+  // Loop over the jets and do cleaning
+  for (uint ijet = 0; ijet < jets->size(); ijet++) {
+    auto const jet = (*jets)[ijet];
+    auto const withinEtaRange = (std::abs(jet.eta()) < jetEtaMax_) and (std::abs(jet.eta()) > jetEtaMin_);
+    // Cuts are applied to jets that are back to back with MET
+    if (std::abs(reco::deltaPhi(jet.phi(), met_phi)) <= dphiJetMetMin_) {
+      cleaned_jets->emplace_back(jet);
+      continue;
+    }
+    // If the jet is outside the range that we're interested in,
+    // append it to the cleaned jets collection and continue
+    if ((jet.pt() < jetPtMin_) or (!withinEtaRange)) {
+      cleaned_jets->emplace_back(jet);
+      continue;
+    }
+
+    JetRef const jetRef(jets, ijet);
+
+    // Cuts related to eta and phi width of HF showers:
+    // Sigma eta-phi based cut + central strip size cut
+    // We'll apply the sigma eta+phi dependent cut only if requested via the config file
+    if (applySigmaEtaPhiCut_) {
+      // Sigma eta eta and sigma phi phi
+      auto const sigmaEtaEta = sigmaEtaEtas[jetRef];
+      auto const sigmaPhiPhi = sigmaPhiPhis[jetRef];
+
+      // Check if sigma eta eta and sigma phi phi are both set to -1.
+      // If this is the case, we can skip the jet without evaluating any mask.
+      if ((sigmaEtaEta < 0) and (sigmaPhiPhi < 0)) {
```

Comment:
Just for completeness, could you elaborate a bit on what can cause these vmaps to return negative values?

Should this plugin also handle the case where only 1 of the 2 is negative? (e.g. `(sigmaEtaEta < 0) or (sigmaPhiPhi < 0)`)

Reply:
Sure, so the vmaps are coming from [this](https://github.com/cms-sw/cmssw/blob/master/RecoJets/JetProducers/plugins/HFJetShowerShape.cc) module. And for `sigmaEtaEta` and `sigmaPhiPhi` values, the placeholder value is `-1`. I believe this happens in two cases:

1. The jet is not within the eta range (default is `|eta|>2.9`)
2. The jet is within the eta range, but it has `pt < ptMin`, `ptMin` specified from the configuration of that module
3. These values might still end up as `-1` due to [here](https://github.com/cms-sw/cmssw/blob/master/RecoJets/JetProducers/plugins/HFJetShowerShape.cc#L181-L187). I believe this is the case where the PU corrected candidate `pt`s are less than a threshold specified from the configuration.

So we shouldn't have to handle the case where only one of them is `-1`, because either both values end up as `-1`, or both result in non-dummy values. Let me know if you have further questions or comments, thanks!

Reply:
Thanks for the clarification. My only suggestion then would be to add a comment somewhere in this plugin saying that the valuemaps are normally created by the producer [HFJetShowerShape](https://github.com/cms-sw/cmssw/blob/master/RecoJets/JetProducers/plugins/HFJetShowerShape.cc), and take this last tweak as a chance to squash in 1 commit.

Reply:
Sure, I've added a comment and squashed everything into one commit in [8d6ce5a](https://github.com/cms-sw/cmssw/pull/37539/commits/8d6ce5aabc6aee7e79282ed98240595eae1e0474).

Reply:
One more question on this.

If `applyStripSizeCut_==true` and the vmap values are negative, the jet is retained in L160 regardless of whether or not it passes the strip-size cut. Is this the intended behaviour? (if the jet fails the strip-size cut, shouldn't it be discarded even if its sigma-eta/phi values are negative?).

Going back to the point on `! ((sigmaEtaEta < 0) and (sigmaPhiPhi < 0))` vs `! ((sigmaEtaEta < 0) or (sigmaPhiPhi < 0))`, I think we should go with the latter, because that (i.e. both positive) is the regime for which the cut "sigma_eta - sigma_phi > X" is meaningful. Effectively it makes no difference now, but this would make this plugin more self-consistent.

In general, we now have in the jet loop several repetitions of the pattern (in pseudo-code)
```cpp
 if (dont_cut_on_this_jet or cut_passes) {
      cleaned_jets->emplace_back(jet);
      continue;
 }
```
It might be clearer to organise this in a different way, e.g.
```cpp
for ( jet_loop ) {
  bool keepJet = true;
  if ( pT > A and B<|eta|<C) {
    if ( abs(deltaPhi) > D) {
      if (applySigmaEtaPhiCut_) {
        if (sigmaEtaEta >= 0 and sigmaPhiPhi >= 0) {
          keepJet &= (sigmaEtaEta - sigmaPhiPhi) <= sigmaEtaPhiDiffMax_;
          if (applySigmaEtaPhiCornerCut_) {
            keepJet &= (sigmaEtaEta > cornerCutSigmaEtaEta_) or (sigmaPhiPhi > cornerCutSigmaPhiPhi_);
          }
        }
      }
      if (keepJet and applyStripSizeCut_) {
        keepJet &= (centralEtaStripSizes[jetRef] <= centralEtaStripSizeMax_);
      }
    }
  }
  if (keepJet) {
    cleaned_jets->emplace_back(jet);
  }
}
```
(Note: this is pseudo-code, and I didn't validate any of this; there might be mistakes)

That said, this point could be addressed in a follow-up PR. This PR could go as is in time for the next pre-release (Tue Apr-19) such that the new plugin becomes available in ConfDB.

Reply:
You're right, but if the vMap values for `sigmaEtaEta` and `sigmaPhiPhi` are negative, I think it is not possible for a jet to fail the `centralEtaStripSize` cut. Referring to the three cases above, if case 1 or 2 is True, we would have the strip size equal to dummy value, `0`. Case 3 is less obvious, but if all PF candidate `pt`s (PU subtracted) are less than the `widthPtThreshold`, they will also be less than the `stripPtThreshold`, since (at least by default values)  `stripPtThreshold > widthPtThreshold`. (The `stripPtThreshold` cutoff is used for counting the strips and adding them to the strip size variable [here](https://github.com/cms-sw/cmssw/blob/master/RecoJets/JetProducers/plugins/HFJetShowerShape.cc#L162-L164).). So the `centralEtaStripSize` for this case would be `0` as well.

Regarding your other comment, I agree that going with `! ((sigmaEtaEta < 0) or (sigmaPhiPhi < 0))` would be more self-consistent, even though it effectively makes no difference now. I also agree that re-organizing the code as you suggested can make things cleaner. But as you pointed out, I can do this re-organization (together with the change in the `if` statement mentioned in the previous sentence) in a follow-up PR if that works. Thanks so much for your review and your comments!

Reply:
Regarding ConfDB and trigger studies, I had an additional question. According to [here](https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideGlobalHLT#CMSSW_12_4_X), the trigger studies are done by pulling the latest `CMSSW_12_3_0/GRun` tables, if I understand correctly. For this plugin to be available on those tables, would I need an additional backport PR to `CMSSW_12_3_X`? Thanks so much for the help!

Reply:
Sure. That works.

Reply:
> Regarding ConfDB and trigger studies, I had an additional question. According to [here](https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideGlobalHLT#CMSSW_12_4_X), the trigger studies are done by pulling the latest `CMSSW_12_3_0/GRun` tables, if I understand correctly. For this plugin to be available on those tables, would I need an additional backport PR to `CMSSW_12_3_X`? Thanks so much for the help!

The answer to the question is 'yes'. On the other hand, I assume these developments are related to [CMSHLT-2282](https://its.cern.ch/jira/browse/CMSHLT-2282), and the latter targets the so-called HLT-V2 menu, which will be in `12_4_X`, so in that sense a backport of this PR to `12_3_X` is not really needed (plus, I would like to avoid 2 backports b/c of the technical tweaks discussed above). The current plan is to move HLT-menus to `12_4_X` after May-2. The normal course of action would be that (1) we integrate this PR now, (2) this producer becomes available in ConfDB after `12_4_0_pre3` is released and you can then use it in your ConfDB areas, and (3) the integration of CMSHLT-2282 will happen once the GRun menu moves to `12_4_X` in early May.

Does this work for you?

Reply:
Yes that's right, these developments are related to the HF-filtered MHTNoMu triggers in [CMSHLT-2282](https://its.cern.ch/jira/browse/CMSHLT-2282), and targets the HLT-V2 menu. I understand, thanks for explaining, I think that should work for me. 

So if I understand correctly, the `12_4_0_pre3` release will be available on Apr-19, so that this module becomes available on ConfDB. I can use it, together with the latest `12_3_0/GRun` tables to finalize my studies on rates, timing and integration tests. And we can integrate these paths on early May when the GRun menu moves to `12_4_X`, does that sound good? Thanks so much again for the help!

Reply:
> So if I understand correctly, the 12_4_0_pre3 release will be available on Apr-19, so that this module becomes available on ConfDB.

Yes.

> I can use it, together with the latest 12_3_0/GRun tables to finalize my studies on rates, timing and integration tests.

In general, yes. This might not be entirely smooth depending on what exactly you want to do. As a last resort for your studies between now and the end of April, you can always add these producers to a local `12_3_X` area of yours (a local manual backport of sorts). In any case, I'm sure one can figure out a proper way to proceed with studies despite these minor complications; you can always ask in the hlt-support channel in case of doubts.

By the way, before the integration in the central tables, the update of these Paths should be discussed/presented in TSG. Or has that already happened?

Reply:
I understand, thanks. Sure then, I can try to do include these producers on my local `12_3_X` area and do the studies as such. I'll contact the Mattermost channel in case of issues, thanks for the help.

I haven't presented this in TSG yet, but I have a talk scheduled on the upcoming meeting, April 20, where I can discuss the cleaning cuts that are applied in this producer, and the checks we did on the impact on noise-enriched events. With this module getting into ConfDB on April 19, I'm not sure if I can get all the studies completed by the next day (rates, timing, integration tests), but would it work if I show some preliminary results I got on Apr 20, and then complete the studies before the end of April? Or would you prefer to postpone the talk until I got those all done, I would be okay with the option that works best for the TSG.

Reply:
Understood. I don't do the TSG schedule :) ; in case of doubts, feel free to ask Darin and Silvio by email.

In any case, the producers don't really need to be in ConfDB to be able to extract preliminary numbers for rates, timing, etc. I think you already have everything right now to put together those numbers (one can do everything in `python` manipulating the existing HLT menus). The details needed for the integration (ConfDB area, etc) can certainly be ironed out after a 1st presentation in TSG.

Reply:
Ah okay got it, thanks! Sure then I can pull existing HLT menus via `hltGetConfiguration`, and do the updates in the python config file to put together the results for Apr 20. And we can iron out the details about the integration after a 1st presentation as you said, thanks for the help!